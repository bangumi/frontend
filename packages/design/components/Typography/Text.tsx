import React, { PropsWithChildren } from 'react'
import classNames from 'classnames'

export interface TextProps {
  className?: string
  style?: React.CSSProperties
  type?: 'default' | 'secondary'
}

const Text = ({
  children,
  className,
  style,
  type = 'default'
}: PropsWithChildren<TextProps>) => {
  return (
    <span
      className={
        classNames(
          'bgm-text', {
            'bgm-text--secondary': type === 'secondary'
          },
          className
        )
      }
      style={style}
    >
      {children}
    </span>
  )
}

export default Text
