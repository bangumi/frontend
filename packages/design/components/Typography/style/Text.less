@text-color-regular: #4b4545;
@text-color-secondary: #aaa6a6;

.bgm-text {
  color: @text-color-regular;

  &--secondary {
    color: @text-color-secondary;
  }
}
