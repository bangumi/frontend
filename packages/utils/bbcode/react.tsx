import React from 'react';

import { convert } from './convert';
import { Parser } from './parser';
import type { CodeNodeTypes, NodeTypes, VNode } from './types';

/* eslint-disable no-useless-escape */
const toCamelCase = (s: string) => s.replace(/(\-\w)/g, (k) => k[1]?.toUpperCase() ?? '');

const joinClassName = (cls: string | string[]) => {
  if (Array.isArray(cls)) {
    return cls.join(' ');
  }
  return cls;
};

const convertToReactProps = (node: VNode, key?: React.Key) => {
  const { props, style, className } = node;

  const transformedStyle: Record<string, string> = style
    ? Object.entries(style).reduce((pre, [key, val]) => {
        return {
          ...pre,
          [toCamelCase(key)]: val,
        };
      }, {})
    : {};

  const transformedClassName = className !== undefined ? joinClassName(className) : undefined;

  const reactProps = {
    ...props,
    style: transformedStyle,
    className: transformedClassName,
    key,
  };

  return reactProps;
};

export const renderNode = (node: NodeTypes, key?: React.Key): React.ReactElement | string => {
  if (typeof node === 'string') {
    const splittedStr: Array<string | React.ReactElement> = node.split(/\n/g);

    if (splittedStr.length < 2) {
      return splittedStr[0] ?? '';
    }
    const children = splittedStr.reduce<Array<string | React.ReactElement>>((pre, curr, idx) => {
      if (idx > 0) {
        return [...pre, React.createElement('br', null), curr];
      }
      return [...pre, curr];
    }, []);
    return React.createElement(React.Fragment, { key }, ...children);
  }

  const { type, children } = node;

  if (children) {
    return React.createElement(
      type,
      convertToReactProps(node, key),
      ...children.map((child, idx) => renderNode(child, idx)),
    );
  }
  return React.createElement(type, convertToReactProps(node, key));
};

export const render = (rawStr: string): Array<React.ReactElement | string> => {
  const nodes: CodeNodeTypes[] = new Parser(rawStr).parse();

  return nodes.map((node, idx) => renderNode(convert(node, {}), idx));
};
