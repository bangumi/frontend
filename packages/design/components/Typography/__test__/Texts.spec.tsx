import React from 'react';
import { render } from '@testing-library/react';
import Text from '../Text';

it.each`
  type
  ${undefined}
  ${'default'}
  ${'secondary'}
`('should render $type text', ({ type }) => {
  const { container } = render(<Text type={type}>hello world</Text>);
  expect(container.firstChild).toMatchSnapshot();
});
