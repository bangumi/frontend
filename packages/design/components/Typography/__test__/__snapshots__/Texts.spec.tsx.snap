// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render default text 1`] = `
<span
  class="bgm-text"
>
  hello world
</span>
`;

exports[`should render secondary text 1`] = `
<span
  class="bgm-text bgm-text--secondary"
>
  hello world
</span>
`;

exports[`should render undefined text 1`] = `
<span
  class="bgm-text"
>
  hello world
</span>
`;
